Add statuspage

jakirkham opened the pull request:
The purpose here is to package [statuspage]( https://github.com/jayfk/statuspage/ ). However, it became clear that statuspage itself was providing binaries. So, in an attempt to build from source, we ended up with things like waf and PyInstaller. These have all either been written from scratch or added via `conda skeleton pypi` and tweaked.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found some lint.

Here's what I've got...


For **recipes/waf**:

 * The recipe must have some tests.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found it was in an excellent condition.

jakirkham commented:
So, it seems that this is challenging to package. Mainly we are having trouble with PyInstaller finding the Python library. Otherwise, it seems waf and PyInstaller work. Though this means we cannot package statuspage from source yet. However, we could package statuspage as a binary. If that is acceptable, we could do this now and focus on building the status page instead. Thoughts?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pycrypto```, ```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found it was in an excellent condition.

jakirkham commented:
Made some progress on building from source. Somehow `six` is missing though. Seems like `setuptools` expects to find it.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/packaging```, ```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found it was in an excellent condition.

jakirkham commented:
Packaging has been broken out in this PR ( https://github.com/conda-forge/staged-recipes/pull/818 ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyinstaller```, ```recipes/python-statuspage```, ```recipes/waf```) and found it was in an excellent condition.

jakirkham commented:
Weird. I cannot reproduce the Travis CI failure. All of this builds fine on my Mac.

jakirkham commented:
Alright, I got this to build successfully on Linux, which is honestly all I'm concerned about ATM. Will add skips as needed to handle packages that don't build on certain platforms. Otherwise, this is ready for review.

jakirkham commented on recipes/python-statuspage/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+{% set version = url_version.replace("-", "") %}
+
+package:
+  name: python-statuspage
```
I prefixed this. I don't what our status is on prefixing these days. That being said, if we decide to try a different status page in the future, I wanted to keep the space open for alternatives.

pelson replied:
:+1:

jakirkham replied:
I dropped the prefix as they have now added a package to PyPI under this name. Not sure how we feel about this. Though it seems like the current trend is away from prefixing.

jakirkham commented on recipes/waf/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+
+test:
+  commands:
+    - waf --help
```
This fails on Windows. I'm totally baffled and am completely uninterested in debugging it ATM. Will raise an issue/PR in the feedstock to explain/demonstrate the problem.

jakirkham commented on recipes/python-statuspage/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+    - click
+    - jinja2
+    - tqdm
+    - six
```
This is a weird package. Basically it wraps all of this into one executable binary. Hence no run-time requirements. However, that is how PyInstaller works.

jakirkham commented on recipes/python-statuspage/meta.yaml:
```diff
@@ -0,0 +1,58 @@
+
+build:
+  number: 0
+  skip: true  # [win or py3k]
```
Had issues getting it to pick up libraries for Python 3.x mainly on Mac (only on Travis CI). As we don't need multiple Python builds, but only need one. I figured it would be simplest to pick the one that worked reliably. Will raise an issue in the feedstock and investigate further later.

jakirkham commented on recipes/pyinstaller/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+    - pyi-set_version = PyInstaller.utils.cliutils.set_version:run
+  script:
+    - source activate "${CONDA_DEFAULT_ENV}"                    # [unix]
+    - export LDFLAGS="${LDFLAGS} -L${PREFIX}/lib"               # [linux]
```
It won't find `libz` without this. Though the other libraries it found find. This was the workaround that seemed to solve the issue so we added it here.

jakirkham replied:
Seems we may have needed this on OS X too. See this issue ( https://github.com/conda-forge/pyinstaller-feedstock/issues/3 ) for details. Fixing in the feedstock with this PR ( https://github.com/conda-forge/pyinstaller-feedstock/pull/4 ).

jakirkham commented:
We're passing. 😄

Builds on Linux and Mac. Currently have skipped Windows due to some issues encountered. As Linux is the primary target at present, I'm not really concerned with getting the others working any time soon.

jakirkham commented:
Should add that PyInstaller does something weird where it generates a bunch of files in a hidden directory `.waf-1.8.21-XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX`. Not really sure whether this should be expected or whether it serves a purpose. If someone knows more, please feel free to share. Otherwise, I'm happy to save that for the feedstock as it doesn't seem to cause any issues.

jakirkham commented:
To get this working, we need to patch PyGithub with an existing upstream PR ( https://github.com/PyGithub/PyGithub/pull/379 ). This can be done easily to our own package for `pygithub` as can be seen with this PR ( https://github.com/conda-forge/pygithub-feedstock/pull/9 ).

jakirkham commented:
Appears that we need to tell the `statuspage` executable how to find the certificates in some cases. I have found that setting `SSL_CERT_FILE` works reasonably well. After giving it some thought, I found the best way to set these variables was have `ca-certificates` set them with activation scripts, which is done in this PR ( https://github.com/conda-forge/ca-certificates-feedstock/pull/2 ). Also, I have tried using that PR in combination with the `statuspage` executable and found that it works nicely.

pelson commented:
Looks fine to me. :+1:

jakirkham commented on recipes/waf/meta.yaml:
```diff
@@ -0,0 +1,39 @@
+{% set name = "waf" %}
```
IDK. This is a generic build tool like `conda-build` that seems to be somewhat well known not really a library that one imports. Do we want to go down this road as it seems to take us into some weird places (e.g. `scons` -> `python-scons`, `salt` -> `python-salt`, `vagrant` -> `ruby-vagrant`, or `mercurial` -> `python-mercurial`)? I like the idea of prefixing generally, but these are cases that start making me go cross-eyed.

jakirkham commented:
So, I have broken out waf into this PR ( https://github.com/conda-forge/staged-recipes/pull/1136 ) and PyInstaller into this PR ( https://github.com/conda-forge/staged-recipes/pull/1137 ). I'm hoping this will let us bank these and thus simplify this PR a bit.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-statuspage```) and found it was in an excellent condition.

jakirkham commented:
As waf and PyInstaller have been broken out and converted into feedstocks, I have removed them from this PR and rebased. This should make it easier to focus on what needs to be done here.

jakirkham commented:
Failures are a product of anaconda.org issues ( https://github.com/Anaconda-Platform/support/issues/49 ). Once those are resolved this will likely pass.

jakirkham commented:
Should note that we are still blocked by the `ca-certificates` activation change ( https://github.com/conda-forge/ca-certificates-feedstock/pull/4 ) and somehow handling additional aspects of the GitHub API via PyGithub ( https://github.com/conda-forge/pygithub-feedstock/pull/9 ). The former PR seems likely to be accepted as it has been previously. However some subtle Windows issues are causing it to fail. The latter is unlikely to get accepted. This makes it incredibly challenging to complete this.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/statuspage```) and found it was in an excellent condition.

jakirkham commented on recipes/statuspage/build.sh:
```diff
@@ -0,0 +1,34 @@
+# Install a development build of PyGithub.
+# This is a workaround for the fact there is no PyGithub release with the changes needed.
+# See PR ( https://github.com/conda-forge/pygithub-feedstock/pull/9 ).
+# Also see PR ( https://github.com/PyGithub/PyGithub/pull/379 ).
```
xref: https://github.com/PyGithub/PyGithub/pull/379

jakirkham commented on recipes/statuspage/build.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Install a development build of PyGithub.
+# This is a workaround for the fact there is no PyGithub release with the changes needed.
+# See PR ( https://github.com/conda-forge/pygithub-feedstock/pull/9 ).
```
xref: https://github.com/conda-forge/pygithub-feedstock/pull/9

jakirkham commented on recipes/statuspage/build.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Install [de]activation scripts for determining where to find the certificates.
+# This is a workaround for the fact that this has not been added to `ca-certificates`.
+# See PR ( https://github.com/conda-forge/ca-certificates-feedstock/pull/4 ).
```
xref: https://github.com/conda-forge/ca-certificates-feedstock/pull/4

jakirkham commented:
As workarounds, I have folded in the patch ( https://github.com/conda-forge/pygithub-feedstock/pull/9 ) that would have been applied to PyGithub into this PR. Don't worry this can live safely with a PyGithub install as we use the PyInstaller provided approach to create one executable that contains everything.

Also, as we have gone the executable route, we needed to set environment variables so that we can find the certificates. However, that PR ( https://github.com/conda-forge/ca-certificates-feedstock/pull/4 ) has not been accepted either. So we have folded in the activation scripts into this recipe for now. They can be removed once the other is addressed.

Given all of these changes, this works fine for me locally and should have no more blockers. However, it is due for a fresh review given these changes. If it is deemed acceptable, please feel free to merge.

jakirkham commented:
Once this is accepted, we should figure out how to add this to Heroku.

jakirkham commented:
This PR ( https://github.com/conda-forge/conda-forge-webservices/pull/34 ) is an "in the works" version of what we need to do, but it will need this package to proceed.

jakirkham commented:
Going to go ahead and self-merge this. Hope that is ok. I'm still interested in getting feedback on it, but it feels like it has sat in staged-recipes too long.
